src/Bicep.Core/Emit/FunctionPlacementValidatorVisitor.cs

```diff
@@ -0,0 +1,146 @@
+                if (popped != element)
+                {
+                    //this error is thrown only if we forgot to implement pop in the visitor or we implement it wrong
+                    throw new InvalidOperationException($"Unexpected element on visited stack. Expecting syntax {element.syntax} with symbol {element.symbol?.Name}, got {popped.syntax} with symbol {popped.symbol?.Name}");
+                }
+
+            }
+        }
+
+        private class VisitorElementScope : IDisposable
+        {
+            private readonly FunctionPlacementValidatorVisitor visitor;
+            private readonly VisitedElement element;
+
+            public VisitorElementScope(FunctionPlacementValidatorVisitor visitor, VisitedElement element)
+            {
+                this.visitor = visitor;
+                this.visitor.visitedElementsStack.Push(element);
+                this.element = element;
+
+            }
+            public void Dispose()
+            {
+                var popped = visitor.visitedElementsStack.Pop();
+                if (popped != element)
+                {
+                    //this error is thrown only if we forgot to implement pop in the visitor or we implement it wrong
+                    throw new InvalidOperationException($"Unexpected element on visited stack. Expecting element {element}, got {popped}");
+                }
+
+            }
+        }
+        private FunctionPlacementValidatorVisitor(SemanticModel semanticModel, IDiagnosticWriter diagnosticWriter)
+        {
+            this.semanticModel = semanticModel;
+            this.diagnosticWriter = diagnosticWriter;
+        }
+
+        public static void Validate(SemanticModel semanticModel, IDiagnosticWriter diagnosticWriter)
+        {
+            var visitor = new FunctionPlacementValidatorVisitor(semanticModel, diagnosticWriter);
+
+            // visiting writes diagnostics in some cases
+            visitor.Visit(semanticModel.SyntaxTree.ProgramSyntax);
+        }
+
+        protected override void VisitInternal(SyntaxBase node)
+        {
+            using var _ = new VisitorSymbolScope(this, node, semanticModel.GetSymbolInfo(node));
+            base.VisitInternal(node);
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            using var _ = new VisitorElementScope(this, VisitedElement.Module);
+            base.VisitModuleDeclarationSyntax(syntax);
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            var vistingModuleParams = visitedElementsStack.Contains(VisitedElement.Module)
+                && !(visitedElementsStack.TryPeek(out var head) && head == VisitedElement.ModuleParams)
+                && syntax.Key is IdentifierSyntax identifierSyntax
+                && string.Equals(identifierSyntax.IdentifierName, "params", StringComparison.OrdinalIgnoreCase);
```

Comment:
>"params" [](http://example.com/codeflow?start=66&length=8)

worth adding a constant to match the name of this with the same property where we calculate the module type?

Reply:
It turned out constant was already there, I used it.